models/cart.py

```diff
@@ -0,0 +1,38 @@
+# cart is a container for unsold items to collect items from different (sub-)categories
+# to be able to checkout this cart at once together with a shipment fee. Only the
+# quantity, category, subcategory is stored because the unique item is not yet sold
+#
+# note that the item is NOT reserved or blocked so that the availability of the item
+# needs to be checked again during checkout
+from typing import List
+
+from sqlalchemy import Column, Integer, ForeignKey, Float
+from sqlalchemy.orm import Mapped, mapped_column, relationship
+
+from models.base import Base
+
+
+class Cart(Base):
+    __tablename__ = "carts"
+
+    id = Column(Integer, primary_key=True)
+    telegram_id: Mapped[int] = mapped_column(ForeignKey("users.telegram_id"))
+    is_closed: Mapped[bool] = mapped_column(default=False)
+
+    shipment: Mapped[int] = mapped_column(default=0, nullable=False)
+    cart_items: Mapped[List["CartItem"]] = relationship(back_populates="cart")
+
+
+class CartItem(Base):
+
+    __tablename__ = "cart_items"
+
+    id = Column(Integer, primary_key=True)
+    cart_id = Column(Integer, ForeignKey("carts.id"), nullable=False)
+    cart: Mapped["Cart"] = relationship(back_populates="cart_items")
+    category_id: Mapped[int] = mapped_column(ForeignKey("categories.id"))
+    category_name: Mapped[str] = mapped_column(default=None, nullable=True)
+    subcategory_id: Mapped[int] = mapped_column(ForeignKey("subcategories.id"), passive_deletes="all")
+    subcategory_name: Mapped[str] = mapped_column(default=None, nullable=True)
+    quantity = Column(Integer, default=0)
+    a_piece_price = Column(Float, default=0.0)
```

Comment:
1) Don't use mapped_column, look at what other models look like and redesign those models to look like them.
2) The CartItem model should be moved to a separate file like I did with BuyItem
3) It is correct to use foreign key only to the primary key of another table, in this case using telegram_id is not correct, even if it works.
4) The cart_items field as I understand it is not even created in the database, and judging by the code I don't quite understand why it is needed. 
5) Why do you store category_name and subcategory_name if it can be obtained via category_id and subcategory_id?
6) The “a_piece_price” field can be retrieved by querying the items table by category_id and subcategory_id.

Reply:
@1: maybe my understanding is wrong, but from what I learnt from [here](https://docs.sqlalchemy.org/en/20/orm/declarative_tables.html#declarative-table-with-mapped-column) is that the way that I'm defining the models not only works the way you defined your models but in a declarative way, plus "mapped_column() supersedes the use of Column()"

@2: true

@3: that made me thinking. First, I agree that per definition an FK is a reference to another table's PK, but as long as the refererred column is unique, from a technical point an FK constraint is okay. But I see your point and want to let you know my idea here: if I would refer the user by its ID, I would have to make another DB query to first get the user.ID from the telegram ID because the telegram ID is already contained in the message object. The message does of course not contain the user.id field. I think that this extra DB query is unneccessary so I wanted to use what's already there. I can redesign it if you want.

@4: I was sticking to the documentation from said documentation of [declarative definition](https://docs.sqlalchemy.org/en/20/orm/declarative_tables.html#declarative-table-with-mapped-column)

@5: to reduce DB queries. From my experience, sometimes the de-normalization of DB schemas is useful when you see an impact on the performance. Let's picture a shop where you would accumulate hundrets of CartItems in one purchase. Since the (sub-)category names are listed in the confirm message, I would have to query the DB once for every CartItem in the Cart. I think that this bit of redundancy would be better for the performance and not increase storage a lot, because the CartItem is meant to be sort of short-lived

@6: same as 5

Again: happy to discuss